torch/csrc/jit/codegen/cuda/scheduler/mma_utils.cpp

```diff
@@ -0,0 +1,422 @@
+      break;
+    case MmaDimension::N:
+      axis_id = layout == MmaOptions::MmaInputLayout::TN ? 1 : 2;
+      break;
+    default:
+      TORCH_INTERNAL_ASSERT(false, "Unreachable");
+      break;
+  }
+
+  int root_size = tv->getMaybeRFactorDomain().size();
+  // Convert to index from right.
+  return tv->getMaybeRFactorDomain().at(root_size + axis_id - 3);
+}
+
+// Locate the root id corresponding to the given mma dimension
+//  Assumes the mma dimension always the innermost 2 or 3, might
+//  need to extend for more complex fusions.
+IterDomain* getMmaOperandRootDimension(
+    TensorView* tv,
+    MmaOptions options,
+    MmaDimension mma_dimension) {
+  if (isVolta(options.macro)) {
+    return getMmaOperandRootDimension3d(
+        tv, options.operand_layout, mma_dimension);
+  }
+  TORCH_INTERNAL_ASSERT(false, "unreachable");
+  return nullptr;
+}
+
+// Preliminary checks to try to validate that leaf is
+//  a innermost dim of root of exactly the given size.
+bool canValidateIsInnerDim(
+    IterDomain* root,
+    IterDomain* leaf,
+    int inner_dim_size) {
+  // Accept boundary case for Volta.
+  if (leaf == root && leaf->isBroadcast()) {
+    return true;
+  }
+  auto expr = leaf->definition();
+  ExpressionEvaluator const_eval(leaf->fusion());
+  auto maybe_leaf_size = const_eval.evaluate(leaf->extent());
+  if (!maybe_leaf_size.has_value()) {
+    return false;
+  }
+  if (maybe_leaf_size.value() != inner_dim_size) {
+    return false;
+  }
+
+  while (expr) {
+    if (auto split = dynamic_cast<Split*>(expr)) {
+      // Inner split only
+      if (leaf != split->inner()) {
+        return false;
+      }
+      // Const split only
+      auto maybe_factor = const_eval.evaluate(split->factor());
+      if (!maybe_factor.has_value()) {
+        return false;
+      }
+      int factor = maybe_factor.value();
+      if (factor < inner_dim_size) {
+        // This might be too restrictive. Would need more
+        //   bookkeeping to relax.
+        return false;
+      }
+      leaf = split->in();
+    } else if (auto merge = dynamic_cast<Merge*>(expr)) {
+      // Might consider just rejecting merge.
+      auto outer = merge->outer();
+      auto inner = merge->inner();
+      if (outer->isBroadcast()) {
+        return false;
+      }
+
+      // Only support merging with constant sized dims
+      maybe_leaf_size = const_eval.evaluate(leaf->extent());
+      if (!maybe_leaf_size.has_value()) {
+        return false;
+      }
+      if (maybe_leaf_size.value() != inner_dim_size) {
+        return false;
+      }
+      leaf = merge->inner();
+    } else {
+      // No support for swizzled inner dim for now.
+      //  Might need to add transpose swizzle here.
+      return false;
+    }
+    expr = leaf->definition();
+  }
+  return leaf == root;
+}
+
+} // namespace
+
+void checkDimSize(
+    TensorView* tv,
+    std::vector<int> axis,
+    std::vector<int> expect) {
```

Comment:
Nit: add some asserts on `axis.size() == expect.size()` and `axis (maybe += nDims) < nDims and >= 0`

Reply:
Added. Thanks.